File: lib/tailmix/scripting/operations/http.rb
# frozen_string_literal: true

module Tailmix
  module Scripting
    module Operations
      module Http
        extend self

        def fetch(interpreter, args)
          options, callback_body = args
          service_class_name = options[:service]

          return unless service_class_name

          handler = service_class_name.safe_constantize
          unless handler&.respond_to?(:call)
            warn "[Tailmix Interpreter Warning] Service '#{service_class_name}' not found or does not respond to .call"
            return
          end

          params = interpreter.eval(options[:params] || {})
          result = handler.call(params, interpreter.server_context)

          response_context = {
            success: result.respond_to?(:success?) ? result.success? : true,
            status: result.try(:status) || 200,
            result: result.try(:data) || result,
            error: result.try(:error)
          }

          interpreter.eval_with_response_context(callback_body, response_context)
        end
      end
    end
  end
end

File: app/javascript/tailmix/runtime/operations/http.js
// Operations for server interaction.
export const HttpOperations = {
    fetch: async (interpreter, args, context) => {
        const [options, callbackBody] = args;
        const { url, method = 'get', params = {}, service } = options;

        const resolvedUrl = await interpreter.eval(url, context);
        const urlWithParams = new URL(resolvedUrl, window.location.origin);
        const fetchOptions = {
            method: method.toUpperCase(),
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json',
            },
        };

        const resolvedParams = {};
        for (const [key, value] of Object.entries(params)) {
            resolvedParams[key] = await interpreter.eval(value, context);
        }

        if (method.toLowerCase() === 'get') {
            Object.entries(resolvedParams).forEach(([key, value]) => {
                urlWithParams.searchParams.append(key, value);
            });
        } else {
            fetchOptions.body = JSON.stringify(resolvedParams);
        }

        let responseContext;
        try {
            const response = await window.fetch(urlWithParams.toString(), fetchOptions);
            const responseData = await response.json();
            responseContext = {
                success: response.ok,
                status: response.status,
                result: responseData,
                error: response.ok ? null : { message: response.statusText },
            };
        } catch (error) {
            console.error("Tailmix fetch error:", error);
            responseContext = {
                success: false,
                status: null,
                result: null,
                error: { message: error.message },
            };
        }

        if (callbackBody) {
            await interpreter.run(callbackBody, responseContext);
        }
    },
};
